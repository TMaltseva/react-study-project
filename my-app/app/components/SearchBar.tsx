import React, { useState, useEffect } from 'react';

interface SearchProps {
  onSearch: (term: string, page: number) => void;
}

const useSearchTerm = () => {
  const [searchTerm, setSearchTerm] = useState(() => {
    if (typeof window !== 'undefined') {
      return localStorage.getItem('searchTerm') || '';
    }
    return '';
  });

  useEffect(() => {
    if (typeof window !== 'undefined') {
      localStorage.setItem('searchTerm', searchTerm);
    }
  }, [searchTerm]);

  return [searchTerm, setSearchTerm] as const;
};

const SearchBar: React.FC<SearchProps> = ({ onSearch }) => {
  const [searchTerm, setSearchTerm] = useSearchTerm();

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearch = () => {
    const trimmedSearchTerm = searchTerm.trim();
    if (typeof window !== 'undefined') {
      localStorage.setItem('searchTerm', trimmedSearchTerm);
    }
    onSearch(trimmedSearchTerm, 1);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      handleSearch();
    }
  };

  return (
    <div>
      <label htmlFor="search-input">Search:</label>
      <input
        id="search-input"
        type="text"
        value={searchTerm}
        onChange={handleInputChange}
        onKeyDown={handleKeyDown}
        aria-label="Search input"
      />
      <button onClick={handleSearch}>Search</button>
    </div>
  );
};

export default SearchBar;
