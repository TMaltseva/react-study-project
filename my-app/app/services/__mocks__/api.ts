export const useFetchDetailsQuery = jest.fn();
