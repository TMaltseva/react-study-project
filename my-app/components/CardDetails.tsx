'use client';

import React from 'react';

interface Details {
  name: string;
  height: string;
  mass: string;
  hair_color: string;
  skin_color: string;
  eye_color: string;
  birth_year: string;
  gender: string;
}

interface CardDetailsProps {
  details: Details;
}

const CardDetails: React.FC<CardDetailsProps> = ({ details }) => {
  return (
    <div className="details">
      <h2>{details.name}</h2>
      <p>Height: {details.height}</p>
      <p>Mass: {details.mass}</p>
      <p>Hair Color: {details.hair_color}</p>
      <p>Skin Color: {details.skin_color}</p>
      <p>Eye Color: {details.eye_color}</p>
      <p>Birth Year: {details.birth_year}</p>
      <p>Gender: {details.gender}</p>
    </div>
  );
};

export default CardDetails;
