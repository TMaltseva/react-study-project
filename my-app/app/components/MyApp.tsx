import React from 'react';
import { Provider } from 'react-redux';
import { AppProps } from 'next/app';
import { wrapper } from '../store/store';
import ThemeProvider from './ThemeProvider';

const MyApp: React.FC<AppProps> = ({ Component, pageProps }) => {
  const { store, props } = wrapper.useWrappedStore(pageProps);

  return (
    <Provider store={store}>
      <ThemeProvider>
        <Component {...props.pageProps} />
      </ThemeProvider>
    </Provider>
  );
};

export default MyApp;
