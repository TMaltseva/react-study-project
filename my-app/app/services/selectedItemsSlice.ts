import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { SearchResult } from '../components/SearchResults';

interface SelectedItemsState {
  items: SearchResult[];
}

const initialState: SelectedItemsState = {
  items: typeof window !== 'undefined' ? JSON.parse(localStorage.getItem('selectedItems') || '[]') : [],
};

const selectedItemsSlice = createSlice({
  name: 'selectedItems',
  initialState,
  reducers: {
    addItem: (state, action: PayloadAction<SearchResult>) => {
      state.items.push(action.payload);
      if (typeof window !== 'undefined') {
        localStorage.setItem('selectedItems', JSON.stringify(state.items));
      }
    },
    removeItem: (state, action: PayloadAction<string>) => {
      state.items = state.items.filter((item) => item.url !== action.payload);
      if (typeof window !== 'undefined') {
        localStorage.setItem('selectedItems', JSON.stringify(state.items));
      }
    },
    clearItems: (state) => {
      state.items = [];
      if (typeof window !== 'undefined') {
        localStorage.setItem('selectedItems', JSON.stringify(state.items));
      }
    },
  },
});

export const { addItem, removeItem, clearItems } = selectedItemsSlice.actions;
export default selectedItemsSlice.reducer;
