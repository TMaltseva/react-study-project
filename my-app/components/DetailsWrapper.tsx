'use client';

import React from 'react';
import { useFetchDetailsQuery } from '../services/api';
import CardDetails from './CardDetails';

interface DetailsWrapperProps {
  id: string;
  onClose: () => void;
}

const DetailsWrapper: React.FC<DetailsWrapperProps> = ({ id, onClose }) => {
  const { data, isLoading, error } = useFetchDetailsQuery(id);

  if (isLoading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <p>Error loading details</p>;
  }

  return (
    <div className="right-section">
      <button onClick={onClose}>Close</button>
      {data && <CardDetails details={data} />}
    </div>
  );
};

export default DetailsWrapper;
