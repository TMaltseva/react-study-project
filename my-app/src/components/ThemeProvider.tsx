import React, { useState, useEffect } from 'react';
import { ThemeContext, Theme } from '../services/themeContext';

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme | null>(null);

  useEffect(() => {
    const savedTheme = (localStorage.getItem('theme') as Theme) || 'light';
    setTheme(savedTheme);
    document.body.className = `${savedTheme}`;
  }, []);

  useEffect(() => {
    if (theme !== null) {
      document.body.className = `${theme}`;
    }
  }, [theme]);

  const toggleTheme = () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    localStorage.setItem('theme', newTheme);
  };

  if (theme === null) {
    return <div>Loading...</div>;
  }

  return <ThemeContext.Provider value={{ theme, toggleTheme }}>{children}</ThemeContext.Provider>;
};

export default ThemeProvider;
