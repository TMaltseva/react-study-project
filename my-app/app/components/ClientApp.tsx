'use client';

import React, { useState } from 'react';
import { useSearchParams, useRouter } from 'next/navigation';
import { useFetchPeopleQuery } from '../services/api';
import SearchBar from './SearchBar';
import SearchResults from './SearchResults';
import Pagination from './Pagination';
import DetailsWrapper from './DetailsWrapper';
import ThemeToggleButton from './ThemeToggleButton';
import Flyout from './Flyout';

const ClientApp: React.FC = () => {
  const searchParams = useSearchParams();
  const router = useRouter();
  const searchTerm = searchParams.get('search') || '';
  const page = parseInt(searchParams.get('page') || '1', 10);
  const { data, isLoading, isFetching } = useFetchPeopleQuery({ searchTerm, page });
  const [selectedId, setSelectedId] = useState<string | null>(null);

  const handleSearch = (term: string, page: number = 1) => {
    const params = new URLSearchParams({ page: page.toString(), search: term });
    router.push(`/?${params.toString()}`);
  };

  const handlePageChange = (page: number) => {
    const params = new URLSearchParams({ page: page.toString(), search: searchTerm });
    router.push(`/?${params.toString()}`);
  };

  const handleItemClick = (id: string) => {
    setSelectedId(id);
    const params = new URLSearchParams({ page: page.toString(), search: searchTerm, details: id });
    router.push(`/?${params.toString()}`);
  };

  const handleCloseDetails = () => {
    setSelectedId(null);
    const params = new URLSearchParams({ page: page.toString(), search: searchTerm });
    router.push(`/?${params.toString()}`);
  };

  return (
    <main className="sections-wrapper">
      <div className="top-section">
        <SearchBar onSearch={(term) => handleSearch(term, 1)} />
        <ThemeToggleButton />
        <Flyout />
      </div>
      <div className="bottom-section">
        <div className="left-section">
          {isLoading || isFetching ? (
            <p>Loading...</p>
          ) : (
            <SearchResults results={data?.results || []} onItemClick={handleItemClick} />
          )}
        </div>
        <div className="right-section">
          {selectedId && <DetailsWrapper id={selectedId} onClose={handleCloseDetails} />}
        </div>
      </div>
      {data?.results?.length ? (
        <Pagination currentPage={page} totalPages={data.totalPages} onPageChange={handlePageChange} />
      ) : null}
    </main>
  );
};

export default ClientApp;
