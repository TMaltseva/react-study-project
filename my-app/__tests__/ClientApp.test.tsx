import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { api } from '../app/services/api';
import selectedItemsReducer from '../app/services/selectedItemsSlice';
import ClientApp from '../app/components/ClientApp';
import { ThemeProvider } from '../app/components/ThemeProvider';
import { useRouter, useSearchParams } from 'next/navigation';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn(),
}));

const mockPush = jest.fn();
(useRouter as jest.Mock).mockReturnValue({ push: mockPush });
(useSearchParams as jest.Mock).mockReturnValue({
  get: jest.fn((key: string) => {
    if (key === 'search') return 'Luke';
    if (key === 'page') return '1';
    return null;
  }),
});

const store = configureStore({
  reducer: {
    [api.reducerPath]: api.reducer,
    selectedItems: selectedItemsReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(api.middleware),
});

describe('ClientApp', () => {
  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ThemeProvider>
          <ClientApp />
        </ThemeProvider>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});