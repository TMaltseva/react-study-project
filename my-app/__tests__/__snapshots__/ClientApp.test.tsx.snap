// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientApp renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="sections-wrapper"
  >
    <div
      class="top-section"
    >
      <div>
        <label
          for="search-input"
        >
          Search:
        </label>
        <input
          aria-label="Search input"
          id="search-input"
          type="text"
          value=""
        />
        <button>
          Search
        </button>
      </div>
      <button>
        Switch to dark
      </button>
      <div
        class="flyout "
      />
    </div>
    <div
      class="bottom-section"
    >
      <div
        class="left-section"
      >
        <p>
          Loading...
        </p>
      </div>
      <div
        class="right-section"
      />
    </div>
  </main>
</DocumentFragment>
`;
